src/transactions/test/InvokeHostFunctionTests.cpp

```diff
@@ -40,294 +40,169 @@
+    auto accAddr = makeAccountAddress(acc.getPublicKey());
 
-        // Make sure sponsorship info hasn't changed
-        {
-            LedgerTxn ltx(app->getLedgerTxnRoot());
-            checkSponsorship(ltx, a1.getPublicKey(), 1, &root.getPublicKey(), 0,
-                             2, 0, 2);
-            checkSponsorship(ltx, root.getPublicKey(), 0, nullptr, 0, 2, 2, 0);
-        }
-    }
+    // Transfer and mint to account
+    test.transfer(root, accAddr, 10, true);
+    test.mint(issuer, accAddr, 10, true);
 
-    SECTION("trustline")
-    {
-        auto const minBalance = app->getLedgerManager().getLastMinBalance(2);
-        auto issuer = root.create("issuer", minBalance);
-        auto acc = root.create("acc", minBalance);
-        auto sponsor = root.create("sponsor", minBalance * 2);
+    // Now mint by transfering from issuer
+    test.transfer(issuer, accAddr, 10, true);
 
-        Asset idr = makeAsset(issuer, "IDR");
-        root.changeTrust(idr, 100);
-        {
-            auto tx = transactionFrameFromOps(
-                app->getNetworkID(), sponsor,
-                {sponsor.op(beginSponsoringFutureReserves(acc)),
-                 acc.op(changeTrust(idr, 100)),
-                 acc.op(endSponsoringFutureReserves())},
-                {acc});
-
-            LedgerTxn ltx(app->getLedgerTxnRoot());
-            TransactionMetaFrame txm(ltx.loadHeader().current().ledgerVersion);
-            REQUIRE(tx->checkValid(*app, ltx, 0, 0, 0));
-            REQUIRE(tx->apply(*app, ltx, txm));
-            REQUIRE(tx->getResultCode() == txSUCCESS);
-            ltx.commit();
-        }
+    // Now burn by transfering to the issuer and by using burn function
+    test.transfer(acc, makeAccountAddress(issuer.getPublicKey()), 10, true);
+    test.burn(acc, 10, true);
 
-        {
-            LedgerTxn ltx(app->getLedgerTxnRoot());
-            auto tlAsset = assetToTrustLineAsset(idr);
-            checkSponsorship(ltx, trustlineKey(acc, tlAsset), 1,
-                             &sponsor.getPublicKey());
-            checkSponsorship(ltx, acc, 0, nullptr, 1, 2, 0, 1);
-            checkSponsorship(ltx, sponsor, 0, nullptr, 0, 2, 1, 0);
-        }
+    // Now transfer and mint to contractAddress
+    auto contractAddr = makeContractAddress(sha256("contract"));
+    test.transfer(root, contractAddr, 10, true);
+    test.mint(issuer, contractAddr, 10, true);
 
-        issuer.pay(root, idr, 100);
+    // Now mint by transfering from issuer
+    test.transfer(issuer, contractAddr, 10, true);
 
-        auto idAndExec = createAssetContract(idr);
+    // Now clawback
+    test.clawback(issuer, accAddr, 2, true);
+    test.clawback(issuer, contractAddr, 2, true);
 
-        auto contractID = idAndExec.first;
-        auto contractExecutableKey = idAndExec.second;
+    // Try to clawback more than available
+    test.clawback(issuer, accAddr, 100, false);
+    test.clawback(issuer, contractAddr, 100, false);
```

Comment:
How about also covering clawback from admin?

Reply:
Do you mean updating the admin to a different account and then clawing back from it? Or clawing back from the issuer?

Reply:
The latter, `clawback(issuer, issuer)`. Actually updating the admin is interesting as well - IIUC `transfer` from non-issuer admin is not the same as `transfer` from issuer (which does `mint` effectively). Or do they behave consistently?

Reply:
Either way, I'd prefer to do any more work in a follow up to avoid bitrot.

Reply:
A `transfer` from a non-issuer admin works like a regular transfer. The issuer is the only one that is special.